sdk/keyvault/keyvault-keys/src/cryptography/remoteCryptographyProvider.ts

```diff
@@ -0,0 +1,390 @@
+  WrapResult,
+  VerifyOptions,
+  VerifyResult,
+  KeyVaultKey,
+  CryptographyClientOptions,
+  DecryptParameters,
+  DecryptOptions,
+  DecryptResult,
+  UnwrapKeyOptions,
+  SignOptions,
+  SignResult,
+  logger,
+  GetKeyOptions
+} from "..";
+import { createSpan } from "../tracing";
+import { challengeBasedAuthenticationPolicy } from "../../../keyvault-common";
+import { SDK_VERSION } from "../constants";
+import { UnwrapResult } from "../cryptographyClientModels";
+import { KeyVaultClient } from "../generated";
+import { parseKeyVaultKeyId } from "../identifier";
+import { LATEST_API_VERSION } from "../keysModels";
+import { getKeyFromKeyBundle } from "../transformations";
+import { createHash } from "./hash";
+import { CryptographyProvider } from "./models";
+import { Span } from "@opentelemetry/api";
+
+/**
+ * The remote cryptography provider is used to run crypto operations against KeyVault.
+ * @internal
+ */
+export class RemoteCryptographyProvider implements CryptographyProvider {
+  constructor(
+    key: string | KeyVaultKey,
+    credential: TokenCredential,
+    pipelineOptions: CryptographyClientOptions = {}
+  ) {
+    const libInfo = `azsdk-js-keyvault-keys/${SDK_VERSION}`;
+
+    const userAgentOptions = pipelineOptions.userAgentOptions;
+
+    pipelineOptions.userAgentOptions = {
+      userAgentPrefix:
+        userAgentOptions && userAgentOptions.userAgentPrefix
+          ? `${userAgentOptions.userAgentPrefix} ${libInfo}`
+          : libInfo
+    };
+
+    const authPolicy = isTokenCredential(credential)
+      ? challengeBasedAuthenticationPolicy(credential)
+      : signingPolicy(credential);
+
+    const internalPipelineOptions = {
+      ...pipelineOptions,
+      loggingOptions: {
+        logger: logger.info,
+        allowedHeaderNames: [
+          "x-ms-keyvault-region",
+          "x-ms-keyvault-network-info",
+          "x-ms-keyvault-service-version"
+        ]
+      }
+    };
+
+    this.client = new KeyVaultClient(
+      pipelineOptions.serviceVersion || LATEST_API_VERSION,
+      createPipelineFromOptions(internalPipelineOptions, authPolicy)
+    );
+
+    this.key = key;
+
+    let keyId: string;
+    if (typeof key === "string") {
+      keyId = key;
+    } else {
+      keyId = key.id!;
+    }
+
+    try {
+      const parsed = parseKeyVaultKeyId(keyId);
+      if (parsed.name === "") {
+        throw new Error("Could not find 'name' of key in key URL");
+      }
+
+      if (!parsed.version || parsed.version === "") {
+        throw new Error("Could not find 'version' of key in key URL");
+      }
+
+      if (!parsed.vaultUrl || parsed.vaultUrl === "") {
+        throw new Error("Could not find 'vaultUrl' of key in key URL");
+      }
+
+      this.vaultUrl = parsed.vaultUrl;
+      this.name = parsed.name;
+      this.version = parsed.version;
+    } catch (err) {
+      logger.error(err);
+
+      throw new Error(`${keyId} is not a valid Key Vault key ID`);
+    }
+  }
+
+  async encrypt(
```

Comment:
Would it make sense to add comments introducing all of these methods?

Reply:
Because these all implement a common interface, and the interface is heavily annotated, you get IntelliSense support with full doc comments here so it felt unnecessary duplication (even though unfortunately it doesn't show up in GH UI) - Let me know if you disagree and I can add them it just felt like duplication and our editor handles the doc comments from the interface well.